xds/internal/testutils/e2e/server.go

```diff
@@ -0,0 +1,135 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package e2e provides utilities for end2end testing of xDS functionality.
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+
+	v3listenerpb "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	v3discoverygrpc "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
+	v3cache "github.com/envoyproxy/go-control-plane/pkg/cache/v3"
+	v3server "github.com/envoyproxy/go-control-plane/pkg/server/v3"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+)
+
+var logger = grpclog.Component("xds-e2e")
+
+// serverLogger implements the Logger interface defined at
+// envoyproxy/go-control-plane/pkg/log. This is passed to the Snapshot cache.
+type serverLogger struct{}
+
+func (l serverLogger) Debugf(format string, args ...interface{}) { logger.Infof(format, args...) }
+func (l serverLogger) Infof(format string, args ...interface{})  { logger.Infof(format, args...) }
+func (l serverLogger) Warnf(format string, args ...interface{})  { logger.Warningf(format, args...) }
+func (l serverLogger) Errorf(format string, args ...interface{}) { logger.Errorf(format, args...) }
+
+// ManagementServer is a thin wrapper around the xDS control plane
+// implementation provided by envoyproxy/go-control-plane.
+type ManagementServer struct {
+	// Address is the host:port on which the management server is listening for
+	// new connections.
+	Address string
+
+	version int
+	gs      *grpc.Server
+	xs      v3server.Server
+	cache   v3cache.SnapshotCache
+}
+
+// StartManagementServer initializes a management server which implements the
+// AggregatedDiscoveryService endpoint. The management server is initialized
+// with no resources. Tests should call the Update() method to change the
+// resource snapshot held by the management server, as required by the test
+// logic. When the test is done, it should call the Stop() method to cleanup
+// resources allocated by the management server.
+func StartManagementServer(ctx context.Context) (*ManagementServer, error) {
+	// Create a snapshot cache.
+	cache := v3cache.NewSnapshotCache(true, v3cache.IDHash{}, serverLogger{})
+	logger.Infof("Created new snapshot cache...")
+
+	// Create an xDS management server and register the ADS implementation provided by it rough a gRPC server.
+	xs := v3server.NewServer(ctx, cache, v3server.CallbackFuncs{})
```

Comment:
This `v3server` can only be "closed" by cancelling this context :man_shrugging:

The `StartManagementServer` function should create a `context.WithCancel`, and cancel it on `Stop`, instead of taking a context from the test.

Reply:
Checking the code of the `v3server` does look like cancelling the context is the only way to get it out of the `process` method. But I tried increasing the `defaultTestTimeout` to `25s`, but still our leak detector was not firing. So, I'm wondering whether there is another code path through which it gets closed.

Either way, I created a context with a cancelFunc based on the context provided to `StartManagementServer` so that we can both honor the test timeout and cancel the `v3server`.

Reply:
`StartManagementServer` no longer needs to take the context as input, right?
The client checking test timeout should be enough. And the returned server is always `defer Stop`ed, so cleanup is also done properly.

Reply:
But if we don't accept a `context` here, we would have to use a `context.Background`. I think accepting a context from the test is better than that.

Reply:
I think it's normal to create context from background, and cancel it to quit the goroutines. We do that in ClientConn, and some of the balancers (though many of them are replaced with `grpcsync.Event` now).

Passing ctx in `StartManagementServer` makes it look like it will block. And also you can cancel the ctx to stop the underlying control plane server, but not the full e2e server. I would still prefer to get rid of it.

Reply:
Done.